libs/SalesforceSDK/src/com/salesforce/androidsdk/security/KeyStoreWrapper.java

```diff
@@ -123,42 +113,107 @@ public synchronized String getRSAPublicString(String name, int length) {
+        }
+        return publicKey;
+    }
+
+    private String getPublicKeyString(String algorithm, String name, int length) {
+        final PublicKey publicKey = getPublicKey(algorithm, name, length);
+        String publicKeyBase64 = null;
+        if (publicKey != null) {
+            publicKeyBase64 = Base64.encodeToString(publicKey.getEncoded(),
+                    Base64.NO_WRAP | Base64.NO_PADDING);
+        }
+        return publicKeyBase64;
+    }
+
+    private PrivateKey getPrivateKey(String algorithm, String name, int length) {
         PrivateKey privateKey = null;
-        createRSAKeysIfNecessary(name, length);
+        createKeysIfNecessary(algorithm, name, length);
         try {
             final KeyStore.Entry entry = keyStore.getEntry(name, null);
             if (entry == null) {
                 return null;
             }
             privateKey = ((KeyStore.PrivateKeyEntry) entry).getPrivateKey();
         } catch (Exception e) {
-            SalesforceSDKLogger.e(TAG, "Could not retrieve RSA private key", e);
+            SalesforceSDKLogger.e(TAG, "Could not retrieve private key", e);
         }
         return privateKey;
     }
 
-    private KeyStore loadKeyStore() throws CertificateException, NoSuchAlgorithmException,
-            IOException, KeyStoreException {
-        final KeyStore keyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
-        keyStore.load(null);
-        return keyStore;
-    }
-
-    private void createRSAKeysIfNecessary(String name, int length) {
+    private synchronized void createKeysIfNecessary(String algorithm, String name, int length) {
         try {
             if (!keyStore.containsAlias(name)) {
 
                 // Generates a new key pair.
                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
-                    final KeyPairGenerator kpg = KeyPairGenerator.getInstance(
-                            KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEYSTORE);
-                    kpg.initialize(new KeyGenParameterSpec.Builder(
+                    final KeyPairGenerator kpg = KeyPairGenerator.getInstance(algorithm, ANDROID_KEYSTORE);
+                    final KeyGenParameterSpec.Builder keyGenParameterSpecBuilder = new KeyGenParameterSpec.Builder(
                             name,
                             KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                             .setKeySize(length)
-                            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_PKCS1)
-                            .build());
+                            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_PKCS1);
+
+                    /*
+                     * TODO: Remove this check once minVersion > 28.
+                     */
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                        try {
+                            keyGenParameterSpecBuilder.setIsStrongBoxBacked(true);
```

Comment:
This is the code that requests a `StrongBox Keymaster` implementation where available. This will give us a trusted standalone hardware where the keys are stored. More details [here](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder.html?hl=en#setIsStrongBoxBacked(boolean)).